### PR TITLE
CURL HTTP headers (using visitors).

### PR DESCRIPTION
**Pull request by sergey-shandar**
This PR is the same as #85 but some functions with memory leaks are removed.

### CODE REVIEW

File: sdk/core/core/inc/az_result.h [hunk 1]
```diff
@@ -53,6 +53,7 @@ AZ_INLINE bool az_succeeded(az_result result) { return (result & AZ_ERROR_FLAG)
 enum {
   AZ_ERROR_ARG = AZ_MAKE_ERROR(AZ_CORE_FACILITY, 1),
   AZ_ERROR_BUFFER_OVERFLOW = AZ_MAKE_ERROR(AZ_CORE_FACILITY, 2),
+  AZ_ERROR_OUT_OF_MEMORY = AZ_MAKE_ERROR(AZ_CORE_FACILITY, 3),
```
Comment (antkmsft) [inline sdk/core/core/inc/az_result.h]:
```suggestion
  AZ_ERROR_NOT_IMPLEMENTED = AZ_MAKE_ERROR(AZ_CORE_FACILITY, 4),
  
```

File: sdk/core/core/inc/az_mock_curl.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#include <stdlib.h>
+
+#include <_az_cfg_prefix.h>
+
+AZ_INLINE az_result az_send_request_impl(
+    az_http_request const * const p_request,
+    az_span * const response,
+    bool allow_allocate) {
+  printf("TO BE IMPLEMENTED, allocating opt: %b", allow_allocate);
+  return AZ_OK;
```
Comment (antkmsft) [inline sdk/core/core/inc/az_mock_curl.h]:
```suggestion
  return AZ_ERROR_NOT_IMPLEMENTED;
```

File: sdk/core/core/inc/az_mock_curl.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <stdlib.h>
+
+#include <_az_cfg_prefix.h>
+
+AZ_INLINE az_result az_send_request_impl(
+    az_http_request const * const p_request,
+    az_span * const response,
+    bool allow_allocate) {
+  printf("TO BE IMPLEMENTED, allocating opt: %b", allow_allocate);
```
Comment (antkmsft) [inline sdk/core/core/inc/az_mock_curl.h]:
```suggestion
```

File: sdk/core/core/CMakeLists.txt [hunk 4]
```diff
@@ -11,17 +11,17 @@ if(NOT CURL_FOUND)
...
-    src/az_assumptions.c
-    src/az_base64.c
-    src/az_json_read.c
-    src/az_http_request.c
-    src/az_pair.c
-    src/az_span.c
-    src/az_uri.c
-    src/az_write_span_iter.c
-    )
+add_library (az_core
```
Comment (antkmsft) [inline sdk/core/core/CMakeLists.txt]:
```suggestion
add_library (
  az_core
```

File: sdk/core/core/CMakeLists.txt [hunk 5]
```diff
@@ -11,17 +11,17 @@ if(NOT CURL_FOUND)
...
+  src/az_assumptions.c
+  src/az_base64.c
+  src/az_curl_adapter.c
+  src/az_http_request.c
+  src/az_json_read.c
+  src/az_pair.c
+  src/az_span.c
+  src/az_span_seq.c
+  src/az_uri.c
+  src/az_write_span_iter.c)
```
Comment (antkmsft) [inline sdk/core/core/CMakeLists.txt]:
```suggestion
  src/az_write_span_iter.c
  )
```

File: sdk/core/core/inc/az_mock_curl.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_MOCK_CURL_H
+#define AZ_MOCK_CURL_H
+
+#include <_az_cfg.h>
```
Comment (antkmsft) [inline sdk/core/core/inc/az_mock_curl.h]:
```suggestion
```

File: sdk/core/core/src/az_span_seq.c [hunk 7]
```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_span_seq.h>
+
+#include <az_write_span_iter.h>
```
Comment (antkmsft) [inline sdk/core/core/src/az_span_seq.c]:
```suggestion
#include <az_str.h>
```

File: sdk/core/core/src/az_span_seq.c [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+#include <az_str.h>
```
Comment (antkmsft) [inline sdk/core/core/src/az_span_seq.c]:
```suggestion
#include <az_write_span_iter>
```

File: sdk/core/core/inc/az_http_request.h [hunk 9]
```diff
@@ -11,7 +11,7 @@
 
 #include <_az_cfg_prefix.h>
 
-// request
+#include <stdlib.h>
```
Comment (antkmsft) [inline sdk/core/core/inc/az_http_request.h]:
```suggestion
#include <stddef.h>
```

Comment (antkmsft) [reply sdk/core/core/inc/az_http_request.h]: Also, you don't need to include az_contract and az_str here

File: sdk/core/core/inc/az_mock_curl.h [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+#include <az_http_request.h>
+
+#include <stdlib.h>
```
Comment (antkmsft) [inline sdk/core/core/inc/az_mock_curl.h]:
```suggestion
```

File: sdk/core/core/src/az_curl_adapter.c [hunk 11]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_curl_adapter.h>
+
```
Comment (antkmsft) [inline sdk/core/core/src/az_curl_adapter.c]:
```suggestion

#include <_az_cfg.h>
```

File: sdk/core/core/test/curl_easy_perform_poc.c [hunk 12]
```diff
@@ -3,39 +3,32 @@
...
-  // clean up
-  fclose(f);
-  curl_easy_cleanup(curl);
-
-  printf("Hello world");
-  return _;
+#include <az_write_span_iter.h>
+#include <az_http_request.h>
+#include <az_pair.h>
+#include <az_json_read.h>
```
Comment (antkmsft) [inline sdk/core/core/test/curl_easy_perform_poc.c]:
```suggestion
#include <az_json_read.h>

#include <_az_cfg.h>
```

File: sdk/core/core/inc/az_curl_adapter.h [hunk 13]
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_CURL_ADAPTER_H
+#define AZ_CURL_ADAPTER_H
+
+#include <_az_cfg.h>
```
Comment (antkmsft) [inline sdk/core/core/inc/az_curl_adapter.h]:
```suggestion
```

File: sdk/core/core/inc/az_http_response.h [hunk 14]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_HTTP_RESPONSE_H
+#define AZ_HTTP_RESPONSE_H
+
+#include <az_span.h>
```
Comment (antkmsft) [inline sdk/core/core/inc/az_http_response.h]:
```suggestion
#include <az_span.h>
#include <az_result.h>
```

File: sdk/core/core/inc/az_http_request.h (hunk 9 above)
Comment (antkmsft) [reply sdk/core/core/inc/az_http_request.h]: And do not include files after _cfg_prefix.h>

File: sdk/core/core/test/curl_easy_perform_poc.c (hunk 12 above)
Comment (antkmsft) [inline sdk/core/core/test/curl_easy_perform_poc.c]: Run formatting, it will sort

Comment (sergey-shandar) [reply sdk/core/core/test/curl_easy_perform_poc.c]: I would strongly discourage to sort headers. This type of "formatting" may change compilation order which can be very important. Please, don't sort includes.

**Comment by sergey-shandar** [comment]
@antkmsft could you remove include sorting, it causes more problems than value.

**Pull request merged**